src/backend/executor/nodeModifyTable.c

```diff
@@ -3654,6 +3709,70 @@ ExecEndModifyTable(ModifyTableState *node)
 	 * shut down subplan
 	 */
 	ExecEndNode(outerPlanState(node));
+
+	/* Notify modified leaf relids to QD */
+	if (GP_ROLE_EXECUTE == Gp_role && node->has_leaf_changed)
+		notify_modified_relations_to_QD(node);
+}
+
+/*
+ * notify_modified_relations_local
+ * For SINGLENODE or we are entry db, update the modified relids on local.
+ * To keep consistent, we set the extend protocol data which will be processed
+ * uniformly later at the end of exetuor run.
+ */
+static void
+notify_modified_relations_local(ModifyTableState *node)
+{
+	Assert(epd);
```

Comment:
when reset epd ?

Reply:
see callers of InitExtendProtocolData().

Reply:
I find list_make1  new list , but no free old list ? Am I right ?

Reply:
When consume subtag list, it's already been cleaned up
```c
bms_free(list_nth(epd->subtagdata, subtag)); /* clean up */
```
And the list pointer which is tiny will be freed with MemoryContext.